Implemented Controller Changes for contributor Admin Dashboard: GSoC M1.6

kshitij01042002 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes part of #17440 
2. This PR does the following: 

**This PR cannot be merged before #18560 gets merged.**

-   Implement Controller layer changes required for Contributor Admin Project. It adds one new handler

                  - ContributorDashboardAdminStatsHandler

- There was some logic error which I discovered in recent debugging so the changes are done in

                 - core/storage/suggestion/gae_models.py

## Essential Checklist

- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...",
followed by a short, clear summary of the changes.
- [x] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [x] The linter/Karma presubmit checks have passed on my local machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
- [x] I have assigned the correct reviewers to this PR (or will leave a
comment with the phrase "@{{reviewer_username}} PTAL" if I don't have
permissions to assign reviewers directly).


## Testing doc (for PRs with Beam jobs that modify production server data)

<!--
If this PR affects production server data, please follow
[these instructions](https://github.com/oppia/oppia/wiki/Testing-jobs-and-other-features-on-production#submitting-a-pr-with-a-new-job-or-feature-that-requires-third-party-api)
and link to the job request doc here.

Otherwise, please delete this section.
-->

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- Make sure your PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

kshitij01042002 committed:
failing tests

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
restored changed file

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
modified controller layer for contributor admin dashboard

kshitij01042002 committed:
duplicate changes from M1.5

kshitij01042002 committed:
pylint errors

kshitij01042002 committed:
pylint errors

kshitij01042002 committed:
mypy errors

kshitij01042002 committed:
mypy errors

kshitij01042002 committed:
mypy errors

kshitij01042002 committed:
mypy errors

kshitij01042002 committed:
added test to backend shards

kshitij01042002 committed:
mypy errors

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.6

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -2214,7 +2214,7 @@ def fetch_page(
         offset: int,
         language_code: str,
-        sort_by: Optional[SortChoices],
+        sort_by: Optional[SortChoices.value],
```
I don't think I have seen this typing before. Can you point me to somewhere else in the codebase where things are typed by {{enum}}.value?

If not, please follow existing codebase conventions, or explain clearly why this is different from all the other cases.

kshitij01042002 replied:
Changed it to be `str`, it's fine if we don't do strict type checking here cause we are doing it in controller. So the string that comes from the frontend will also be one of `CD_ADMIN_STATS_SORT_OPTIONS` defined in constants.ts

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -2289,34 +2289,49 @@ def fetch_page(
-                        sorted_results.append(result_model)
-                        if len(sorted_results) == page_size:
+                        filtered_results.append(result_model)
+                        if len(filtered_results) == page_size + offset:
```
Is using `offset` rather than `temp_offset` here correct?

Also please name the "temp_offset" something different. I cannot tell, from the name, how it differs from `offset`. Try and convey the purpose of each variable clearly through its naming.

kshitij01042002 replied:
Yess, the loop should break when we get the (page_size + offset) that is the number of results till the end of desired page.

kshitij01042002 replied:
What other can I name it because it is the temporary offset I am using to skip the results when I am making the request to the datastore again.

seanlip replied:
What is the difference between "temp_offset" and "offset"? Your responses don't explain this clearly.

Also please see my comment below, starting the fetch from item 0 is not acceptable from a latency perspective.

kshitij01042002 replied:
As we have discarded the new approach and keeping the same approach,temp_offset is not used anymore.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -2289,34 +2289,49 @@ def fetch_page(
+                        filtered_results.append(result_model)
+                        if len(filtered_results) == page_size + offset:
                             break
+            sorted_results = filtered_results[offset:]
```
This seems odd to me. Presumably the offset is to skip the results that have already been sent to the frontend. So why are we recomputing them again here? That seems too inefficient to be workable.

kshitij01042002 replied:
The reason we have to compute them is because the last_activity filter is not directly applied in the gae query. We need to fetch the results again and discard the one that has already been sent to frontend.

seanlip replied:
But that won't work due to the inefficiency I mentioned. Why can't you memorize the previous offset and then start counting from that when making the next query?

kshitij01042002 replied:
Yeah, that was exactly what I did the first time I implemented it but while testing I mistakenly took the number of models to skip as same as offset and then created a lot of confusion which resulted in unneccessary efforts.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -2289,34 +2289,49 @@ def fetch_page(
-            TranslationSubmitterTotalContributionStatsModel] = (
-                sort_query.fetch(offset=next_offset))
-        more: bool = len(next_result_model) != 0
+            # Check whether we have more results.
```
If logic is repeated in if/else branches, find a way to pull it out of both branches so that it only needs to be stated once.

kshitij01042002 replied:
Sounds reasonable, updated PTAL.

seanlip commented on core/storage/suggestion/gae_models.py:
```diff
@@ -2636,7 +2651,7 @@ def fetch_page(
         offset: int,
         language_code: str,
-        sort_by: Optional[SortChoices],
+        sort_by: Optional[SortChoices.value],
```
Same question about the use of "value".

kshitij01042002 replied:
Added the reasoning above. PTAL.

seanlip commented on core/storage/suggestion/gae_models_test.py:
```diff
@@ -3459,27 +3459,26 @@ def test_fetch_page_with_filtering(self) -> None:
                 language_code='es'
             ))
-        self.assertEqual(len(sorted_results), 1)
+        # self.assertEqual(len(sorted_results), 1)
```
Why are there comments in your code? Please read through your "Files Changed" tab and make sure the changes are exactly what you want them to be before you ask reviewers to do the same.

kshitij01042002 replied:
Removed it.

seanlip commented on core/storage/suggestion/gae_models_test.py:
```diff
@@ -3459,27 +3459,26 @@ def test_fetch_page_with_filtering(self) -> None:
                 language_code='es'
             ))
-        self.assertEqual(len(sorted_results), 2)
+        self.assertEqual(len(sorted_results), 1)
```
Why are these results changing? Normally we expect tests to be stable when the code changes if the behaviour is still supposed to be the same -- in fact that's one of the reasons we write tests in the first place. So the existence of changes in the expectations here, without any description/discussion of behaviour change, is a red flag.

kshitij01042002 replied:
I have changed the size of the page from 2 to 1, to test more branching and combinations. So the the results are also changing. PTAL above.

seanlip reviewed (changes requested):
@kshitij01042002 Took a pass at codeowner files. I have some concerns about the changes to the storage layer, I think they need more explanation. Please see the inline comments.

Thanks!

kshitij01042002 committed:
changed typings for suggestion/gae_models

kshitij01042002 committed:
Merge branch 'gsoc_M1.6' of https://github.com/kshitij01042002/oppia into gsoc_M1.6

kshitij01042002 committed:
pylint errors

kshitij01042002 committed:
mypy errors

kshitij01042002 committed:
mypy errors

kshitij01042002 committed:
pylint errors

kshitij01042002 committed:
removed duplicate logic

kshitij01042002 commented:
Hi @seanlip I have tried to answer all your comments. PTAL.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.6

DubeySandeep commented on core/controllers/contributor_dashboard_admin.py:
```diff
@@ -529,3 +530,191 @@ def _get_complete_translation_contribution_stats(
+        ]:
+            raise self.InvalidInputException(
+                'Invalid contribution subtype %s.' % (contribution_subtype)
+            )
```
This should be handled via the `URL_PATH_ARGS_SCHEMAS`

kshitij01042002 replied:
Yeah, removed it from here and added in schema validation.

DubeySandeep commented on core/controllers/contributor_dashboard_admin.py:
```diff
@@ -529,3 +530,191 @@ def _get_complete_translation_contribution_stats(
+                'Invalid contribution subtype %s.' % (contribution_subtype)
+            )
+
+        assert self.normalized_payload is not None
```
Do we need this, I'm assuming the `HANDLER_ARGS_SCHEMAS` must be handling it.

kshitij01042002 replied:
Actually this is required for mypy checking. Even though we are validating it in schema it assumes that it can be None.

DubeySandeep commented on core/controllers/contributor_dashboard_admin.py:
```diff
@@ -529,3 +530,191 @@ def _get_complete_translation_contribution_stats(
+        num_days_since_last_activity = self.normalized_payload[
+            'num_days_since_last_activity']
+
+        if contribution_type == feconf.CONTRIBUTION_TYPE_TRANSLATION:
```
I see repeated patterns in all the if-else blocks, can we simplify it:
```
stats, next_offset, more = get_total_stats(contribution_type, contribution_subtype) 
self.render_json({
    'frontend_dicts': [
        stat.to_frontend_dict() for stat in translation_submitter_stats],
    'next_offset': next_offset,
    'more': more
})
```

You can further simplify  `get_total_stats` with a static map `contribution_type => contribution_subtype => method`. (Though this will make the structure bit rigid, if you think that the general structure would be same for all the cases in future then we can use the dict/map.)

kshitij01042002 replied:
I'm little confused like how would we implement static mapping, do we have any existing example in the codebase? 

The function and number of params are different for each case.

DubeySandeep commented on core/controllers/contributor_dashboard_admin.py:
```diff
@@ -529,3 +530,191 @@ def _get_complete_translation_contribution_stats(
+        'POST': {
+            'page_size': {
+                'schema': {
+                    'type': 'int'
```
Do we have positiveInt (or Non-negaitve) as a type?

kshitij01042002 replied:
Not sure, I used the convention already used in the codebase

kshitij01042002 replied:
Added a validator which checks that value should be atleast 0 i.e. a nonNegative

DubeySandeep commented on core/controllers/contributor_dashboard_admin.py:
```diff
@@ -529,3 +530,191 @@ def _get_complete_translation_contribution_stats(
+            },
+            'offset': {
+                'schema': {
+                    'type': 'int'
```
Ditto*

kshitij01042002 replied:
Added a validator which checks that value should be atleast 0 i.e. a nonNegative

DubeySandeep commented on core/controllers/contributor_dashboard_admin.py:
```diff
@@ -529,3 +530,191 @@ def _get_complete_translation_contribution_stats(
+            },
+            'language_code': {
+                'schema': {
+                    'type': 'basestring'
```
Can we use the validator:
```
'validators': [{
    'id': 'is_supported_audio_language_code'
}]
```

kshitij01042002 replied:
Sg. Updated.

DubeySandeep commented on core/controllers/contributor_dashboard_admin.py:
```diff
@@ -529,3 +530,191 @@ def _get_complete_translation_contribution_stats(
+                'schema': {
+                    'type': 'basestring'
+                },
+                'default_value': None
```
is `language_code` an optional param? I don't see it to be optional in `ContributorDashboardAdminStatsHandlerNormalizedPayloadDict`

kshitij01042002 replied:
Actually it's optional cause we don't need language_code for question stats

kshitij01042002 replied:
Thanks for pointing out.

DubeySandeep commented on main.py:
```diff
@@ -290,6 +290,10 @@ def get_redirect_route(
         contributor_dashboard_admin.ContributorDashboardAdminPage),
+    get_redirect_route(
+        r'%s/<contribution_type>/<contribution_subtype>' % (
+            feconf.CONTRIBUTOR_DASHBOARD_ADMIN_STATS_URL),
```
```suggestion
            feconf.CONTRIBUTOR_DASHBOARD_ADMIN_STATS_URL_PREFIX),
```

kshitij01042002 replied:
Changed it. Thanks.

DubeySandeep commented on core/controllers/contributor_dashboard_admin.py:
```diff
@@ -529,3 +530,191 @@ def _get_complete_translation_contribution_stats(
+    GET_HANDLER_ERROR_RETURN_TYPE = feconf.HANDLER_TYPE_JSON
+    URL_PATH_ARGS_SCHEMAS = {
+        'contribution_type': {
+            'schema': {
+                'type': 'basestring'
+            }
```
Maybe add choices here?

kshitij01042002 replied:
I have added a check for choices inside the post method, so that we can through a customized error message.

DubeySandeep replied:
The schema validation for payloads/urls exists so that we don't have to write these validations manually, please use schema validation for any incoming data.

kshitij01042002 replied:
Added the schema validation and removed manual validation. PTAL.

DubeySandeep commented on core/controllers/contributor_dashboard_admin.py:
```diff
@@ -529,3 +530,191 @@ def _get_complete_translation_contribution_stats(
+        },
+        'contribution_subtype': {
+            'schema': {
+                'type': 'basestring'
```
Add choices here*

kshitij01042002 replied:
Ditto above, added check inside the post method.

DubeySandeep replied:
Please add the basic validations via schema. Note: we are not expecting any of these issues to happen vai fronted, it is only possible if someone tries to access these APIs manually and I think ` customized error message.` is not required for such cases.

kshitij01042002 replied:
Sounds reasonable. Updated the checks. PTAL. Thanks

DubeySandeep commented on core/controllers/contributor_dashboard_admin.py:
```diff
@@ -529,3 +530,191 @@ def _get_complete_translation_contribution_stats(
+                    'type': 'basestring'
+                },
+                'default_value': None,
+                'choices': constants.CD_ADMIN_STATS_SORT_OPTIONS
```
Are we expecting choices to be inside `schema`?

kshitij01042002 replied:
I looked out some other schemas and then decided to have a choices validator here.

DubeySandeep replied:
I'm not sure whether this is an expected pattern, can you please add a few links where we have added choices this way? (It will help me understand the situation and maybe fix other places as well if they are not working fine!)

kshitij01042002 replied:
One of the example is [here](https://github.com/oppia/oppia/blob/develop/core/controllers/contributor_dashboard.py#L878).

DubeySandeep commented on core/controllers/contributor_dashboard_admin.py:
```diff
@@ -529,3 +530,191 @@ def _get_complete_translation_contribution_stats(
+            },
+            'num_days_since_last_activity': {
+                'schema': {
+                    'type': 'int'
```
Maybe nonNegative int? (if we have one!)

kshitij01042002 replied:
I am not sure if we have anything specific for Non-negative but we can use validator to have min-value to be Zero.

`
                'validators': [{
                    'id': 'is_at_least',
                    'min_value': 0
                }],
`

Something like this.

DubeySandeep commented on core/controllers/contributor_dashboard_admin.py:
```diff
@@ -529,3 +530,191 @@ def _get_complete_translation_contribution_stats(
+                translation_submitter_frontend_dicts = [stat.to_frontend_dict()
+                    for stat in translation_submitter_stats]
+                response = {
+                    'frontend_dicts': translation_submitter_frontend_dicts,
```
Should we call it `stats` instead on `frontend_dicts`

sagangwee replied:
+1.

kshitij01042002 replied:
Updated. PTAL.

DubeySandeep commented on core/controllers/contributor_dashboard_admin.py:
```diff
@@ -529,3 +530,191 @@ def _get_complete_translation_contribution_stats(
+    }
+
+    @acl_decorators.can_access_contributor_dashboard_admin_page
+    def post(
```
Why is this handled via "post"? are we expecting to update any field in the datastore through this API?

sagangwee replied:
+1. This should be GET.

kshitij01042002 replied:
Actually I took the reference of [this](https://github.com/oppia/oppia/blob/develop/core/controllers/topics_and_skills_dashboard.py#L256) code used for getting data for skills dashboard. Since we are implementing pagination in the backend itself we need to send data via this API. I am not sure about this though, can you please give me suggestion whether should I change it. If yes I'll have to make the changes and also change the tests so will resolve other comments after getting this one resolved.

DubeySandeep replied:
Yes change it to GET as it's a request to fetch data, there are a lot of issues with using post in place of GET. Also, thanks for reporting the [issue](https://github.com/oppia/oppia/blob/develop/core/controllers/topics_and_skills_dashboard.py#L256) in develop., I'll report it to the LaCE team and get it fixed.

DubeySandeep replied:
@kshitij01042002 Is this resolved?

kshitij01042002 replied:
Yeah I have changed the method from POST to GET.

DubeySandeep commented on core/domain/contributor_admin_dashboard_services.py:
```diff
@@ -0,0 +1,458 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Funtions to get stats Displayed in contributor Admin Dashboard.
```
We don't name services files based on the pages instead entity names are used like exp_services, story_services. I would suggest either moving these functions to suggestion_services (As all the models are stored in suggestion_models.py) or renaming it to suggestion_stats_services/contribution_stats_services.py

DubeySandeep replied:
@kshitij01042002 is this resolved?

kshitij01042002 replied:
Yeah. I have renamed the files to contribution_stats_services.py

DubeySandeep reviewed (changes requested):
@kshitij01042002 I've reviewed the controller file and left a few comments, PTAL! (I was going through the changes to know/understand the new features being added to the admin dashboard. The overall changes look great and I'm excited to see the stats on the admin page soon, thanks for working on it. :))

lkbhitesh07 commented on core/feconf.py:
```diff
@@ -922,6 +922,7 @@ def get_empty_ratings() -> Dict[str, int]:
 CONTRIBUTOR_ALL_STATS_SUMMARIES_URL = '/contributorallstatssummaries'
 CONTRIBUTOR_CERTIFICATE_URL = '/contributorcertificate'
 CONTRIBUTOR_DASHBOARD_ADMIN_URL = '/contributor-dashboard-admin'
+CONTRIBUTOR_DASHBOARD_ADMIN_STATS_URL = '/contributor-dashboard-admin-stats'
```
Can you please once confirm the naming convention we use here, at some places we separate the keywords by a hyphen(-), at some places by an underscore(_), and at some places its continuous text, maybe with @seanlip?
Just want to be sure so that we follow a similar pattern.
Thanks

kshitij01042002 replied:
Hi @seanlip , does the convention used correct? Or it should be something like hitesh suggested?

seanlip replied:
User-facing URLs (that are shown in the URL bar) should be hyphenated. Data handlers can be concatenated (but hyphenated is probably OK too).

lkbhitesh07 reviewed (approved):
Thanks for the changes @kshitij01042002,
Approving since the core/domain/contributor_admin_dashboard_services.py and core/domain/contributor_admin_dashboard_services_test.py is part of 18560 and have already been approved there. For the other 2 codeowner files LGTM, also left a comment.
Thanks

seanlip commented on core/domain/contributor_admin_dashboard_services.py:
```diff
@@ -233,7 +233,7 @@ def get_translation_submitter_total_stats(
     offset: int,
     language_code: str,
-    sort_by: Optional[suggestion_models.SortChoices.value],
+    sort_by: Optional[str],
```
Re https://github.com/oppia/oppia/pull/18576#discussion_r1249593166 -- I'm not sure I buy the rationale. The type checking should always be as strict as possible, so I wouldn't suggest relaxing it.

My original comment was about something different. Here it is for reference:

> I don't think I have seen this typing before. Can you point me to somewhere else in the codebase where things are typed by {{enum}}.value?
>
> If not, please follow existing codebase conventions, or explain clearly why this is different from all the other cases.

Can you address the question here?

kshitij01042002 replied:
I am not able to find an example where we have used {enum}.values as a typing but it is necessary here because we have created a new constant CD_ADMIN_STATS_SORT_OPTIONS in constants.ts which is mapped to the values of this enum. On using the constant value directly as typing there is a mypy error

error: Name 'constants.CD_ADMIN_STATS_SORT_OPTIONS.value' is not defined  [name-defined]

seanlip commented on core/domain/contributor_admin_dashboard_services.py:
```diff
@@ -233,7 +233,7 @@ def get_translation_submitter_total_stats(
     offset: int,
     language_code: str,
-    sort_by: Optional[suggestion_models.SortChoices.value],
+    sort_by: Optional[str],
```
Re https://github.com/oppia/oppia/pull/18576#discussion_r1249593166 -- I'm not sure I buy the rationale. The type checking should always be as strict as possible, so I wouldn't suggest relaxing it.

My original comment was about something different. Here it is for reference:

> I don't think I have seen this typing before. Can you point me to somewhere else in the codebase where things are typed by {{enum}}.value?
>
> If not, please follow existing codebase conventions, or explain clearly why this is different from all the other cases.

Can you address the question here?

kshitij01042002 replied:
I am not able to find an example where we have used {enum}.values as a typing but it is necessary here because we have created a new constant `CD_ADMIN_STATS_SORT_OPTIONS` in constants.ts which is mapped to the values of this enum. On using the constant value directly as typing there is a mypy error

`
error: Name 'constants.CD_ADMIN_STATS_SORT_OPTIONS.value' is not defined  [name-defined]
`

seanlip commented on core/storage/suggestion/gae_models_test.py:
```diff
@@ -3459,7 +3459,6 @@ def test_fetch_page_with_filtering(self) -> None:
                 num_days_since_last_activity=7,
                 language_code='es'
             ))
-        # self.assertEqual(len(sorted_results), 1)
```
Why are you removing this assertion?

kshitij01042002 replied:
Reveted the changes.

seanlip commented on core/storage/suggestion/gae_models_test.py:
```diff
@@ -3459,27 +3459,25 @@ def test_fetch_page_with_filtering(self) -> None:
                 num_days_since_last_activity=7,
                 language_code='es'
             ))
-        self.assertEqual(len(sorted_results), 1)
```
Why are you removing this check? Please don't remove existing expectations -- you do need to test the number of results and each result value.

kshitij01042002 replied:
Reverted the changes

sagangwee commented on core/controllers/contributor_dashboard_admin_test.py:
```diff
@@ -763,3 +763,831 @@ def test_stats_data_not_calculated_without_topic_id_and_language_code(
+        self.assertEqual(
+            response['error'], 'Invalid contribution type invalid.')
+
+        self.logout()
```
```suggestion
        csrf_token = self.get_new_csrf_token()
        
        response = self.post_json(
            '/contributor-dashboard-admin-stats/invalid/submission', {
                'page_size': 2,
                'offset': 1,
                'language_code': self.SUGGESTION_LANGUAGE_CODE,
                'sort_by': None,
                'topic_ids': None,
                'num_days_since_last_activity': None
            },
                csrf_token=csrf_token,
                expected_status_int=400)

        self.assertEqual(
            response['error'], 'Invalid contribution type invalid.')
        self.logout()
```

Use new lines to differentiate the test structure (setup, action under test, asserts). Ditto below. Also see https://github.com/oppia/oppia/wiki/Backend-tests#guidelines-for-writing-good-tests.

kshitij01042002 replied:
Updated. PTAL. Thanks for pointing to the documentation.

sagangwee commented on core/controllers/contributor_dashboard_admin.py:
```diff
@@ -529,3 +530,191 @@ def _get_complete_translation_contribution_stats(
+    language_code: str
+    sort_by: Optional[str]
+    topic_ids: Optional[List[str]]
+    num_days_since_last_activity: Optional[int]
```
```suggestion
    max_days_since_last_contribution: Optional[int]
```

Seems clearer.

kshitij01042002 replied:
I think `max_days_since_last_acitivity` would be accurate cause we are also considering for reviewers.

I am thinking to make this change at the end as this will also change it's naming in other files which will result in more reviewers getting assgined. It is just a variable name change so thinking of doing it at the end. 

Is it fine?

sagangwee replied:
Sure, `max_days_since_last_activity` seems reasonable. 

The end of the GSoC period? I would just do it now, as I don't really see the point of deferring it to later.

kshitij01042002 replied:
No not at the end of GSoC at the end of this PR, after all other major comments have been resolved.

sagangwee replied:
I still don't see the benefit of waiting—any added reviewers will need to review the same changes at some point or another. The PR is almost ready anyway, so I would just make the changes.

kshitij01042002 replied:
Done.

sagangwee commented on core/controllers/contributor_dashboard_admin_test.py:
```diff
@@ -763,3 +763,831 @@ def test_stats_data_not_calculated_without_topic_id_and_language_code(
+        self.login(self.CONTRIBUTOR_EMAIL)
+        csrf_token = self.get_new_csrf_token()
+
+        # Test with language filter and pagination.
```
Let's create a separate test for each test case (and create a helper function to setup the models). That way, if a case fails, it's obvious which case is failing.

sagangwee replied:
This doesn't seem addressed?

kshitij01042002 replied:
I have split the test cases into different tests. PTAL.

sagangwee commented on core/storage/suggestion/gae_models.py:
```diff
@@ -2289,34 +2289,45 @@ def fetch_page(
+            TranslationSubmitterTotalContributionStatsModel] = []
+        check_offset: int
 
         if num_days_since_last_activity is not None:
```
Why do you have to filter the results manually? Can you not add a `last_contribution_date` filter to the storage fetch call? e.g. https://cloud.google.com/appengine/docs/legacy/standard/python/ndb/queries#filter_by_prop.

kshitij01042002 replied:
The first sort property must be the same as the property to which an inequality filter is applied. Thus, the inequality filter on last_activity can not be used here and we have implemented it separately below. Learn more about this here:
https://cloud.google.com/appengine/docs/legacy/standard/go111/datastore/query-restrictions#properties_used_in_inequality_filters_must_be_sorted_first.

sagangwee replied:
Ah, right. Forgot about that quirk. Can you please add a comment explaining this with a link to the documentation for future readers?

kshitij01042002 replied:
I have added it above on line 2273 where we are actually defining the sort query. Sean suggested that it would be better to reference it while defining the sort query itself rather than later.

sagangwee reviewed (changes requested):
@kshitij01042002 Thanks took a pass.

kshitij01042002 commented:
Hi @DubeySandeep can you please take a look at this comment https://github.com/oppia/oppia/pull/18576#discussion_r1251996012 and suggest me how to proceed? Thanks.

cc / @sagangwee

DubeySandeep commented:
@kshitij01042002 I've replied to your comment, PTAL!

aasiffaizal commented on core/controllers/contributor_dashboard_admin.py:
```diff
@@ -529,3 +530,191 @@ def _get_complete_translation_contribution_stats(
+    ) -> None:
+        """Handles POST requests."""
+
+        if contribution_type not in [
+            feconf.CONTRIBUTION_TYPE_TRANSLATION,
+            feconf.CONTRIBUTION_TYPE_QUESTION
+        ]:
```
```suggestion
        if contribution_type not in {
            feconf.CONTRIBUTION_TYPE_TRANSLATION,
            feconf.CONTRIBUTION_TYPE_QUESTION
        }:
```

aasiffaizal replied:
also can't we leverage schema validation for these checks?

aasiffaizal commented on core/controllers/contributor_dashboard_admin.py:
```diff
@@ -529,3 +530,191 @@ def _get_complete_translation_contribution_stats(
+            raise self.InvalidInputException(
+                'Invalid contribution type %s.' % (contribution_type)
+            )
+        if contribution_subtype not in [
+            feconf.CONTRIBUTION_SUBTYPE_SUBMISSION,
+            feconf.CONTRIBUTION_SUBTYPE_REVIEW
+        ]:
```
```suggestion
        if contribution_subtype not in {
            feconf.CONTRIBUTION_SUBTYPE_SUBMISSION,
            feconf.CONTRIBUTION_SUBTYPE_REVIEW
        }:
```

aasiffaizal commented on core/storage/suggestion/gae_models.py:
```diff
@@ -2214,7 +2214,7 @@ def fetch_page(
         offset: int,
         language_code: str,
-        sort_by: Optional[SortChoices],
+        sort_by: Optional[str],
```
Why change it from enum to str?

kshitij01042002 replied:
Actually while trying to address https://github.com/oppia/oppia/pull/18576#discussion_r1249593166 this. I changed it from enum to str but we shouldn't be loose on type checking therefore added it back again.

aasiffaizal reviewed (changes requested):
Thanks @kshitij01042002. Took a pass, left a few comments

kshitij01042002 committed:
changed file name

kshitij01042002 committed:
Update core/controllers/contributor_dashboard_admin.py

Co-authored-by: Aasif Faizal <38973423+aasiffaizal@users.noreply.github.com>

sagangwee reviewed (changes requested):
@kshitij01042002 Please, resolve all comments before sending back for review, thanks.

kshitij01042002 committed:
changed typings to constant

kshitij01042002 committed:
Merge branch 'gsoc_M1.6' of https://github.com/kshitij01042002/oppia into gsoc_M1.6

kshitij01042002 committed:
changed typings to constant

kshitij01042002 committed:
pylint errors

kshitij01042002 committed:
reverted the changes which wasn't necessary

kshitij01042002 commented:
Hi @seanlip I have addressed your comments and made some new changes. The previous changes were not necessary it happend due to my misunderstandings and I apologise for the inconvienence caused in the process. If you still have concerns please let me know. Thanks.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.6

kshitij01042002 committed:
Update core/controllers/contributor_dashboard_admin.py

Co-authored-by: Aasif Faizal <38973423+aasiffaizal@users.noreply.github.com>

kshitij01042002 commented:
Hi reviewers, I will fix the failing tests once I have successfully migrated the handler from POST to GET.

gp201 reviewed (approved):
LGTM, I have similar comments as @sagangwee . I'm approving to unblock.  If you make any additional modifications to the test files, please ask @U8NWXD to review them.

seanlip reviewed (commented):
You haven't answered all my review comments, e.g. this and others:

https://github.com/oppia/oppia/pull/18576#discussion_r1253258030

Please reply to all comments before asking for a review. Otherwise reviewers will just reassign your PR back to you, and this will result in delays getting your PR merged. See the PR guidelines on the wiki for more details.

kshitij01042002 commented:
Hi @seanlip sorry for the mis communication actually I replied to the same comment below so didn't replied there. Addressed both the comments. PTAL.

qinghaoyang reviewed (approved):
@kshitij01042002 LGTM.

seanlip commented:
@kshitij01042002 You still have not replied to all review comments (from me + other reviewers), so I am reassigning this PR to you.

Before you assign a PR to any reviewer, please make sure that **all** previous comments in the thread have been fully replied to. Otherwise, reviewers will just assign the PR back to you and ask that you respond to those. Thanks.

kshitij01042002 commented:
Hi @seanlip I have addressed all your comments if there is anyone that I missed to address please let me know. Addressing all the reviews at once is getting quite confusing and time consuming so trying to resolve comments by one reviewer at a time or atleast reviewers of separate files. As you are codeowner of storage files, I asked a review from you on gae_models files.

For handler level changes I am still facing some issues and trying to do the correct changes for this I am working with Qinghao and Sandeep. Once I have done that I'll ask for review on controller files from other reviewers.

Hopefully I have addressed all your concers, if there's still any please let me know.

seanlip reviewed (approved):
Thanks, no further concerns for my codeowner files. I am a bit confused about whether your enum values thing is the final solution or not, but what you have now is probably fine IMO.

Re the unreplied comments, I was looking at Sandeep's, and you also have one missing by @lkbhitesh07: https://github.com/oppia/oppia/pull/18576#discussion_r1252496982.

kshitij01042002 committed:
changed POST to GET

kshitij01042002 committed:
splitted the conditions in different tests

kshitij01042002 committed:
added validator for num_day_since_last_activity

kshitij01042002 committed:
changed name from frontend_dicts to stats

kshitij01042002 committed:
resolved merge conflicts

kshitij01042002 committed:
fixed pylint errors

kshitij01042002 committed:
fixed mypy checks

kshitij01042002 commented:
Hi @seanlip assigning you in concern with https://github.com/oppia/oppia/pull/18576#discussion_r1252496982. Need validaion on that. Thanks.

kshitij01042002 commented:
Hi @DubeySandeep and @sagangwee . I have addressed all your comments. PTAL.

Thanks for the review.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.6

DubeySandeep commented:
@kshitij01042002 Can you please reply to all the comments?

kshitij01042002 commented:
Hi @DubeySandeep I think I have replied to all the comments but in case if I have missed any, please let me know. Thanks.

kshitij01042002 committed:
minor fixes in test file

kshitij01042002 committed:
Merge branch 'gsoc_M1.6' of https://github.com/kshitij01042002/oppia into gsoc_M1.6

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.6

sagangwee commented on core/controllers/contributor_dashboard_admin_test.py:
```diff
@@ -763,3 +764,843 @@ def test_stats_data_not_calculated_without_topic_id_and_language_code(
+        self.logout()
+
+    def test_get_translation_submitter_stats_for_pagination(self) -> None:
+
```
Nit: Remove new line. Ditto below.

kshitij01042002 replied:
Removed.

sagangwee commented on core/controllers/contributor_dashboard_admin_test.py:
```diff
@@ -763,3 +764,843 @@ def test_stats_data_not_calculated_without_topic_id_and_language_code(
+        )
+        self.logout()
+
+    def test_get_translation_reviewer_stats_for_last_activity_filter(self) -> None: # pylint: disable=line-too-long
```
Don't think the lint exception is necessary if we indent properly.
```suggestion
    def test_get_translation_reviewer_stats_for_last_activity_filter(
      self
    ) -> None:
```

kshitij01042002 replied:
Yeah, Thanks for the suggestion.

sagangwee commented on core/domain/contribution_stats_services.py:
```diff
@@ -0,0 +1,458 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Funtions to get stats Displayed in contributor Admin Dashboard.
```
I believe this file will encompass more than just getters.

```suggestion
"""Service methods for operating on contributor dashboard user stats.
```

kshitij01042002 replied:
Changed it to `Service methods for operating on contributor admin dashboard user stats.`. I think it is more accurate as we are using this file for admin dashboard stats.

sagangwee reviewed (changes requested):
Thanks, took a pass. Just some minor comments left.

kshitij01042002 committed:
Update core/domain/contribution_stats_services.py

Co-authored-by: Sagang Wee <9004726+sagangwee@users.noreply.github.com>

kshitij01042002 committed:
minor nits suggested by Sagang

kshitij01042002 committed:
fixed merge conflicts

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.6

kshitij01042002 commented:
Hi @sagangwee Updated the changes you requested. PTAL. Thanks for the review.

sagangwee commented on core/controllers/contributor_dashboard_admin.py:
```diff
@@ -529,3 +530,215 @@ def _get_complete_translation_contribution_stats(
+            'topic_ids': {
+                'schema': {
+                    'type': 'custom',
+                    'obj_type': 'JsonEncodedInString'
```
This seems like a hack that isn't quite what we want since we shouldn't accept non-list like JSON strings. Did you try the suggestion of adding the `topic_ids` URL query param in the URL of `get_json` calls in the tests? 

A `list` should be a valid schema type (https://github.com/oppia/oppia/wiki/Schemas#write-a-schema), so if the above doesn't work, then you can try reaching out to the schema wiki writers: https://github.com/oppia/oppia/wiki/Schemas#write-a-schema. @rohitkatlaa @vojtechjelinek @Nik-09

kshitij01042002 replied:
Actually in the codebase where we have to pass list through a GET method we are using this technique. See for example **ConceptCardDataHandler**.

On trying the method you suggested the error still persist `{"error": "Schema validation for 'topic_ids' failed: Expected list, received topic1", "status_code": 400}`

seanlip replied:
@kshitij01042002 Just a note, I imagine this discussion would have come up when the PR for ConceptCardDataHandler was created. Maybe find that PR using git's blame functionality and see if anything was mentioned there.

kshitij01042002 replied:
The schema was introduced in this PR https://github.com/oppia/oppia/pull/14563 but there isn't any discussion about this.

seanlip replied:
~~OK, I know what happened here -- @sagangwee is right.~~

~~JsonEncodedInString was introduced in https://github.com/oppia/oppia/commit/aefe95731d353fe7142f8b18343e49c81e39adcf because it was necessary to transfer an encoded object in a user-facing URL.~~

~~Its use in data handlers was never actually intended, because you can transfer structured data through a JSON request and #14563 is actually wrong.~~

~~So I would suggest using the appropriate schema for a list of strings instead. We have a wiki page on schemas that you might find useful: https://github.com/oppia/oppia/wiki/Schemas~~

EDIT: Ah, sorry, ignore what I said above. The distinction is not actually between page/data handlers as I mistakenly thought, it is between GET and POST/PUT/etc. handlers. See @Nik-09's comment below.

kshitij01042002 replied:
Using List in schema definition doesn't work as expected for GET requests. It only considers the first element of the list, as mentioned in the error above.

seanlip replied:
Could you dig a little deeper into why it doesn't work? Propagating a bad pattern isn't really the way to go here.

(You can use this [debugging doc template](https://docs.google.com/document/d/1qRbvKjJ0A7NPVK8g6XJNISMx_6BuepoCL7F2eIfrGqM/edit) to organize your investigation.)

Nik-09 replied:
Adding a note: The data passed through URL query params are string data types and in order to normalize the string into a list of topicIDs, here `JsonEncodedInString` is used. Thus @kshitij01042002 is facing the error i.e., "expected list received ...".

kshitij01042002 replied:
Hi @Nik-09 thanks for rolling in. I tried to dig more into it and found the similar conclusion. The URL query doesn't allow to pass the list but only string values. If we want to pass a list either we have to use the `POST` method or pass it as a `JsonEncodedInString`

@seanlip @sagangwee How should I proceed with it?

cc: @DubeySandeep

seanlip replied:
Sorry, I amended my comment above -- I think what you have is probably fine.

sagangwee replied:
I think our validation logic doesn't support URL query params with the same key, though the `webapp2` framework does with `getall`: https://webapp2.readthedocs.io/en/latest/guide/request.html#get-data. So, what we have here seems fine.

sagangwee reviewed (changes requested):
@kshitij01042002 A couple of remaining comments.

kshitij01042002 committed:
changed num_days_since_last_activity to max_days_since_last_activity

kshitij01042002 committed:
Merge branch 'gsoc_M1.6' of https://github.com/kshitij01042002/oppia into gsoc_M1.6

kshitij01042002 commented:
Hi @sagangwee I have addressed your comments PTAL. Thanks.

DubeySandeep commented on core/controllers/contributor_dashboard_admin.py:
```diff
@@ -529,3 +530,215 @@ def _get_complete_translation_contribution_stats(
+                    'type': 'basestring'
+                },
+                'default_value': None,
+                'validators': [{
```
I was expecting this to be inside `schema`, can you please check whether it's working here?

@Nik-09 Can we add validators here? (OR do we need to it inside the schema)? If both, I would suggest sticking it to one place i.e, inside the schema.

Nik-09 replied:
Yes we need to add **choices** inside the **schema**, for ref: https://github.com/oppia/oppia/blob/3d1d9af3d08d8448af866cbe5e5779e4bd9feb90/core/schema_utils.py#L278

kshitij01042002 replied:
Okay. Updated it. PTAL. Thanks.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.6

kshitij01042002 committed:
added validation inside schema

kshitij01042002 commented:
Hi @aasiffaizal @sagangwee @DubeySandeep 

I have addressed all your comments, can you PTAL? Thanks for the review.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.6

kshitij01042002 committed:
removed accidently added code

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.6

sagangwee reviewed (approved):
@kshitij01042002 Thanks, LGTM.

aasiffaizal reviewed (approved):
LGTM
